first draft of clean_column_names utility function

jburos opened the pull request:
Added a function `strip_column_names` that converts column names to patsy-friendly strings.

**Note** the name has changed to `strip_column_names` (it was `clean_column_names`) in an attempt to be less confusing. Clean could be confused with presentation-ready column names.

## Example usage : 

```
from cohorts.utils import strip_column_names
cohort = cohorts.init_cohort(...)
d = cohort.get_dataframe()
d = d.rename(columns = strip_column_names(d.columns))
```

You can now also do the above in one step, as:

```
d = cohort.get_dataframe(rename_cols=True)
```
  (this latter functionality is `off` by default, so the current behavior is unchanged)

## Behavior 

In both cases, the function modifies column names as : 

    1. convert meaningful punctuation to str (e.g. `<` becomes `lt`)
    2. (optionally - not default) remove text within parens
    3. convert remaining punctuation/whitespace -> _
    4. change to lowercase
    5. strip leading & trailing punctuation

It will additionally check for uniqueness of stripped column names, and will return original values if the revised names are non-unique.

## Examples 

Some examples of how this maps from old to new names:

```
{'2-factor score': '2_factor_score',
 '5-factor score': '5_factor_score',
 'Age': 'age',
 'Albumin < 4': 'albumin_lt_4',
 'Alive Status': 'alive_status',
 'Baseline neutrophil to lymphocyte ratio': 'baseline_neutrophil_to_lymphocyte_ratio',
 'Best Response RECIST 1.1': 'best_response_recist_1_1',
 'Best response mRECIST': 'best_response_mrecist',
 'Date of Best Response': 'date_of_best_response',
 'ECOG >0': 'ecog_gt_0',
 'Hb < 10': 'hb_lt_10',
 'Liver Mets': 'liver_mets',
 'Number of Lines of Therapy': 'number_of_lines_of_therapy',
 'OS in days': 'os_in_days',
 'Ongoing Responder RECIST 1.1': 'ongoing_responder_recist_1_1',
 'PD-L1': 'pd_l1',
 'PFS (RECIST 1.1) in days': 'pfs_recist_1_1_in_days',
 'PFS (mRECIST 1.1) in days': 'pfs_mrecist_1_1_in_days',
 'Pack Years': 'pack_years',
 'Prior BCG': 'prior_bcg',
 'Response Mixed or Homogenous': 'response_mixed_or_homogenous',
 'Sequence Path': 'sequence_path',
```

tavinathanson commented on cohorts/utils.py:
```diff
@@ -0,0 +1,124 @@
+import re 
+
+def _clean_column_name(x, keep_parens = True):
+    """
```
Style: lines should be indented at the same level as the `"""`: https://www.python.org/dev/peps/pep-0008/#documentation-strings

tavinathanson commented on cohorts/utils.py:
```diff
@@ -0,0 +1,124 @@
+
+import re 
+
+def _clean_column_name(x, keep_parens = True):
```
Better name than `x`?

tavinathanson commented on cohorts/utils.py:
```diff
@@ -0,0 +1,124 @@
+
+        Parameters 
+        -----------
+        str (str): input character string
```
`str` vs. `x`

tavinathanson commented on cohorts/utils.py:
```diff
@@ -0,0 +1,124 @@
+        Parameters 
+        -----------
+        str (str): input character string
+        keep_parens (logical): control behavior of within-paren elements of text
```
Maybe `keep_paren_contents` would be more accurate?

tavinathanson commented on cohorts/utils.py:
```diff
@@ -0,0 +1,124 @@
+        --------
+        modified string for new field name 
+
+
```
Unnecessary newline

tavinathanson commented on cohorts/utils.py:
```diff
@@ -0,0 +1,124 @@
+
+def clean_column_names(cols, keep_parens = True):
+    """
+        Utility script for renaming pandas columns to patsy-friendly names.
```
Same comment re indentation

tavinathanson commented on test/test_utils.py:
```diff
@@ -0,0 +1,39 @@
+    ## should not rename columns -- should error
+    d3 = d.rename(columns = clean_column_names(d.columns, keep_parens = False))
+    ok_((d3.columns == d.columns).all())
+
```
Excess newlines

tavinathanson commented on test/test_utils.py:
```diff
@@ -0,0 +1,39 @@
+      'PD L1 (value)': pd.Series([1., 2., 3., 4.], index=['a', 'b', 'c', 'd']),
+      'PD L1 (>1)': pd.Series([0., 1., 1., 0.], index=['a', 'b', 'c', 'd']),
+      }
+    d = pd.DataFrame(d)
```
Very minor: might be good to use `df` vs. `d`, seems to be the standard way to name a `DataFrame` e.g. http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html

tavinathanson commented on cohorts/utils.py:
```diff
@@ -0,0 +1,124 @@
+    
+    ## confirm no duplicates in remaining columns
+    if not(len(newcols) == len(cols)):
+        print('Warning: length of newcols not the same as cols')
```
Worth raising an error here?

jburos replied:
Perhaps .. was thinking about how to make this more friendly for interactive use. Is there a way to raise a warning?

tavinathanson replied:
I think the standard warning protocol is via `logging`, which I assume is generally preferable over `print` (and `cohorts` could do better at that as a whole). So I guess the question here is whether the issue is worth halting completely over. I'll defer to you on that call?

tavinathanson replied:
Oops, I think `warnings.warn("warning message", TypeOfWarning)` might be even better? I don't know too much about this, sadly!

tavinathanson replied:
Ignore everything I said! I guess `warning.warn` would raise an exception depending on the current settings for warning suppression.

tavinathanson commented on cohorts/utils.py:
```diff
@@ -0,0 +1,124 @@
+    
+    ## replace meaningful punctuation with text equivalents
+    ## surround each with whitespace to enforce consistent use of _
+    newx = re.sub('<=', ' le ', newx)
```
Maybe avoid the repeated `re.sub` e.g.

```
punctuation_to_text = {"=<": "le", ...}
for punctuation, punctuation_text in punctuation_to_text.items():
    col_name = col_name.replace(punctuation, punctuation_text)
```

jburos replied:
I like this - then the punct-to-string can be more easily extended. Thanks!

tavinathanson commented on cohorts/utils.py:
```diff
@@ -0,0 +1,124 @@
+    newx = re.sub(punct_pattern, punct_replacement, newx)
+    
+    ## remove trailing _ if it exists (if last char was punctuation)
+    newx = re.sub('_$', '', newx)
```
Maybe `newx.strip("_")` is cleaner?

jburos replied:
Would this also strip leading _ ?

tavinathanson replied:
It would, is that a problem?

jburos replied:
Maybe - in some cases, I have used a leading underscore to include information about the version of data in that row. But otherwise I agree with you it would _usually_ be good to strip this.

I will implement strip for now & we'll cross the leading-underscore problem if/when we encounter it

tavinathanson commented on cohorts/utils.py:
```diff
@@ -0,0 +1,124 @@
+        
+        Parameters 
+        --------------
+            cols (list): list of strings containing column names
```
I think this is actually two indentations too far in: should be lined up with `Parameters`, which should be lined up with `"""`

tavinathanson commented on cohorts/utils.py:
```diff
@@ -0,0 +1,124 @@
+    newcols = [_clean_column_name(col, keep_parens = keep_parens) for col in cols]
+    
+    ## confirm no duplicates in remaining columns
+    if not(len(newcols) == len(cols)):
```
When would this be not true? Versus `len(set(newcols))`?

jburos replied:
This is a sanity check - one checks for uniqueness the other checks whether the list-comprehension worked properly

tavinathanson replied:
Two thoughts: `assert` is good for when we want to test basic things that shouldn't actually go wrong. That said, I don't think we need to take up space testing that Python list comprehensions work?

jburos replied:
True. I guess I was thinking if there was an None value or something , but this probably wouldn't catch that anyways.

tavinathanson commented on cohorts/utils.py:
```diff
@@ -0,0 +1,124 @@
+
+    """
+    ## clean/replace punctuation
+    newcols = [_clean_column_name(col, keep_parens = keep_parens) for col in cols]
```
Minor style: `new_cols` vs. `newcols`

tavinathanson commented on cohorts/utils.py:
```diff
@@ -0,0 +1,124 @@
+        print('Warning: length of newcols not the same as cols')
+        return dict(zip(cols, cols))
+    
+    if not(len(newcols) == len(set(newcols))):
```
Why is `lower`ing not just a part of `_clean_column_name`?

tavinathanson commented on cohorts/utils.py:
```diff
@@ -0,0 +1,124 @@
+        return dict(zip(cols, cols))
+    
+    ## modify all column names to lower case
+    newcols_lc = [col.lower() for col in newcols]
```
Why is `lower`ing not just a part of `_clean_column_name`?

jburos replied:
A few reasons:
1. So this can be turned off without having to pass parameter to `_clean_column_names`
2. Separating steps would, I thought, be somewhat cleaner.
3. So uniqueness could be checked before & after case-conversion. Ideally I want to have the non-uniqueness message be more helpful, like "after step X these columns were equivalent - please fix & try again". & actually list the duplicate columns.

tavinathanson replied:
I think checking uniqueness at every step is probably over-engineering this; if a uniqueness error occurs, I think it should be enough to spit out which columns are duplicates, regardless of what step made them so?

I think, readability-wise, it's more intuitive to have the `_clean_column_name` function do everything to each individual column name, while the `clean_column_names` function iterates over them. That would be my assumption, and any deviation seems like added confusion?

Not clear what 1 means: how do you turn it off?

jburos replied:
I haven't added that parameter, but conceivably a user might not want to convert case

tavinathanson commented:
LGTM! % minor style comments

tavinathanson commented:
@jburos also just occurred to me: why manual `d.rename(columns = clean_column_names(d.columns))` vs. `cohort.as_dataframe(rename_cols=True, keep_paren_contents=True)`, with `True` by default?

jburos commented:
@tavinathanson I would suggest we _also_ support that usage. 

I would recommend we keep the functional use separate since this utility 
1. could be used outside of cohorts
2. can be applied to a subset of column names easily this way, and 
3. including it in the as_data_frame call could make errors converting columns (ie string handling resulting in non-unique column names) difficult to address

But, I do agree that this additional functionality could be helpful -- e.g. when you want to apply coxph model to outcome vars & this is only exposed via the `Cohort` object.

Curious to know what you think.

tavinathanson commented:
@jburos I agree that this should be accessible outside `as_dataframe`, but `as_dataframe` seems like a great place to use it. Not too worried about errors in the conversion; if we clearly `warn` when they come up, that seems fine?

jburos commented:
@tavinathanson I modified this in two places : 
1. Renamed function (& helper function) to `strip_column_names`.
2. Added option for `rename_cols` to `as_dataframe` (defaults to False).

Would you mind reviewing these two changes again? 

Thanks & apologies for the CR-overload...

jburos commented on cohorts/load.py:
```diff
@@ -397,14 +412,17 @@ def is_lambda(func):
+
+        if (rename_cols):
+            df = df.rename(columns=_strip_column_names(d.columns,
+                   keep_paren_contents=keep_paren_contents))
```
@tavinathanson wasn't sure here what I should do with `cols` - welcome your advice on this part

tavinathanson commented on cohorts/load.py:
```diff
@@ -328,7 +329,8 @@ def _as_dataframe_unmodified(self, join_with = None, join_how = None):
 
-    def as_dataframe(self, on=None, col=None, join_with=None, join_how=None, **kwargs):
+    def as_dataframe(self, on=None, col=None, join_with=None, join_how=None,
+                    rename_cols=False, keep_paren_contents=True, **kwargs):
```
Style nit: `rename_cols` should be directly underneath the paren

tavinathanson commented on cohorts/load.py:
```diff
@@ -346,6 +348,19 @@ def as_dataframe(self, on=None, col=None, join_with=None, join_how=None, **kwarg
+        Other parameters
+        ----------------
+        `rename_cols`: (bool)
+                if True, then return columns using "stripped" column names
```
Indent by 4 spaces vs. 8? https://www.python.org/dev/peps/pep-0008/#indentation

tavinathanson commented on cohorts/load.py:
```diff
@@ -397,14 +412,17 @@ def is_lambda(func):
-            return (cols, df)
+
+        if (rename_cols):
+            df = df.rename(columns=_strip_column_names(df.columns,
```
I apologize for the confusing nature of this code, but I think this will break in some instances when the name of the function, e.g. `missense_snv_counT` gets stripped here, however it doesn't get stripped in the `cols` variable.

Maybe, instead of putting this `if (rename_cols)` here, you should put it inside `apply_func` above, replacing `df[col] = df.apply(func, axis=1)`?

Something to replace both the column name and the dataframe e.g.:

```
applied = df.apply(func, axis=1)
if (rename_cols):
    df.rename(...)
    col = _strip_column_name(...)
df[col] = applied
```

tavinathanson replied:
Have a simpler thought elsewhere; please ignore this comment!

tavinathanson commented on cohorts/load.py:
```diff
@@ -397,14 +412,16 @@ def is_lambda(func):
-            return (cols, df)
+
+        if (rename_cols):
+            df = df.rename(columns=_strip_column_names(df.columns, keep_paren_contents=keep_paren_contents))
```
Sorry for the confusing nature of this code! I think the `DataFrame` itself might get column-renamed while the `cols` variable won't, which could lead to issues.

I think `cols = _strip_column_names(cols, keep_paren_contents=keep_paren_contents)` would address that?

tavinathanson replied:
Also, you can do an in-place rename with `df.rename(inplace=True)` to avoid `df = df.rename`

jburos replied:
Thanks - was wondering what to do about `cols` -- this will be a bit tricky to do cleanly since I won't know at this point if the column-rename was successful... feel like I should look these up in the dictionary returned by the original _strip_column_names on the df. 

I'll try to write up a test case for this.

tavinathanson replied:
I can't think of a scenario where the rename would fail on a column name in the `DataFrame` that it would not fail on in the `cols`? I agree that this feels a little hacky, and if you want to do it more cleanly feel free; fine with me either way.

tavinathanson commented on cohorts/utils.py:
```diff
@@ -0,0 +1,136 @@
+    # remove leading/trailing _ if it exists (if last char was punctuation)
+    new_col_name = new_col_name.strip("_")
+
+    # TODO: check for empty string
```
Not sure if this is intentional, but seems fine to leave the `TODO` if you make an issue for it

tavinathanson commented on cohorts/utils.py:
```diff
@@ -0,0 +1,136 @@
+    Returns
+    -------
+
+    dict mapping cols -> newcols
```
`new_col`

tavinathanson commented on cohorts/utils.py:
```diff
@@ -0,0 +1,136 @@
+    if not(len(new_cols) == len(set(new_cols))):
+        logger.warning('Warning: strip_column_names (if run) would introduce duplicate names.'
+                       ' Reverting column names to the original.')
+        print('Warning: strip_column_names would introduce duplicate names. Please fix & try again.')
```
Is the reason you're `print`ing in addition to logging because it's not showing up in the notebook? I was reading more of this to try to understand how this should work: http://stackoverflow.com/questions/18786912/get-output-from-the-logging-module-in-ipython-notebook; but I admit that I'm fairly naive and this is fine for now until we get more clarity.

jburos replied:
Yes - that's why. I also ended up adding a `warning.warn` since the `logger.warning` message didn't show up in either the console or the ipynb -- appears this takes a separate setup which I was reading about here : https://docs.python.org/2.6/library/logging.html#configuring-logging

Definitely doesn't belong in this utils module, though should probably be addressed at some point .. for now I have found duplicating these messages helpful.

tavinathanson commented on cohorts/utils.py:
```diff
@@ -0,0 +1,136 @@
+        _strip_column_name(col, keep_paren_contents=keep_paren_contents)
+        for col in cols]
+
+    if not(len(new_cols) == len(set(new_cols))):
```
`len(new_cols) != len(set(new_cols))`?

tavinathanson commented:
@jburos looks good! Once you address the remaining comments, feel free to merge without another review pass (though happy to give more feedback if useful)
